http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java

```diff
@@ -58,31 +62,89 @@ protected void channelRead0(ChannelHandlerContext channelContext, HttpObject msg
+        if (msg instanceof HttpResponse) {
+            HttpResponse response = (HttpResponse) msg;
             SdkHttpResponse sdkResponse = SdkHttpFullResponse.builder()
                                                              .headers(fromNettyHeaders(response.headers()))
                                                              .statusCode(response.status().code())
                                                              .statusText(response.status().reasonPhrase())
                                                              .build();
             channelContext.channel().attr(KEEP_ALIVE).set(HttpUtil.isKeepAlive(response));
             requestContext.handler().headersReceived(sdkResponse);
-            requestContext.handler().onStream(new PublisherAdapter(response, channelContext, requestContext));
+        }
+
+        if (msg instanceof StreamedHttpResponse) {
+            requestContext.handler().onStream(new PublisherAdapter((StreamedHttpResponse) msg, channelContext, requestContext));
+        } else if (msg instanceof FullHttpResponse) {
+            requestContext.handler().onStream(new Publisher<ByteBuffer>() {
+                @Override
+                public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+                    subscriber.onNext(((FullHttpResponse) msg).content().nioBuffer());
+                    channelContext.channel().attr(ChannelAttributeKeys.SUBSCRIBER_KEY)
+                                  .set(subscriber);
+                }
+            });
+        }
+
+        if (msg instanceof LastHttpContent) {
+            Subscriber<? super ByteBuffer> subscriber = channelContext.channel().attr(ChannelAttributeKeys.SUBSCRIBER_KEY).get();
+            try {
+                subscriber.onComplete();
+                requestContext.handler().complete();
+            } catch (RuntimeException e) {
+                subscriber.onError(e);
+                requestContext.handler().exceptionOccurred(e);
+                throw e;
+            } finally {
+                finalizeRequest(requestContext, channelContext);
+            }
+        }
+    }
+
+    private static void finalizeRequest(RequestContext requestContext, ChannelHandlerContext channelContext) {
+        runAndLogError("Could not remove request specific handlers from pipeline",
+                       () -> removePerRequestHandlers(channelContext));
+        if (!channelContext.channel().attr(KEEP_ALIVE).get()) {
+            closeAndRelease(channelContext);
+        } else {
+            requestContext.channelPool().release(channelContext.channel());
+        }
+    }
+
+    private static void removePerRequestHandlers(ChannelHandlerContext ctx) {
+        removePerRequestHandlers(ctx, HttpStreamsClientHandler.class);
+    }
+
+    @SafeVarargs
+    private static void removePerRequestHandlers(ChannelHandlerContext ctx, Class<? extends ChannelHandler>... handlerClasses) {
```

Comment:
Does this need to be varargs anymore? I only see one use. It could even be inlined?

Reply:
Nah I can inline it. Copied it from my other branch which removes some timeout handlers but didn't port that over.